libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -0,0 +1,714 @@
+      file_ending.remove_suffix(extension.size());
+      return kvp;
+    }
+  }
+  return {};
+}
+
+auto find_formatter_given(std::string_view what, auto func, auto member,
+                          std::string_view path, std::string_view compression,
+                          location loc, const char* docs,
+                          session ctx) -> pair_for<decltype(func)> {
+  auto possibilities = std::vector<std::string>{};
+  auto res = find_given(what, func, member, possibilities);
+  if (res.first) {
+    return res;
+  }
+  std::ranges::sort(possibilities);
+  if (loc.end - loc.begin == path.size() + 2) {
+    auto file_start = path.find(what);
+    loc.begin += file_start + 1;
+    loc.end -= 1;
+    loc.end -= compression.size();
+  }
+  auto diag = diagnostic::error("no known format for extension `{}`", what)
+                .primary(loc)
+                .note("supported extensions for format deduction:  `{}`",
+                      fmt::join(possibilities, "`, `"));
+
+  if (compression.empty()) {
+    diag = std::move(diag).note(
+      "supported extensions for compression deduction: `{}`",
+      fmt::join(extension_to_compression_map | std::views::keys, "`, `"));
+  }
+  diag = std::move(diag)
+           .hint("you can pass a pipeline to handle compression and format")
+           .docs(docs);
+  std::move(diag).emit(ctx);
+  return {};
+}
+
+auto strip_scheme(ast::expression& expr, std::string_view scheme) -> void {
+  auto arg = try_as<ast::constant>(expr);
+  TENZIR_ASSERT(arg);
+  auto strip_size = scheme.size() + 3;
+  // remove the quotes and scheme from the location
+  arg->source.begin += 1 + strip_size;
+  // remove the quotes from the location
+  arg->source.end -= 1;
+  match(
+    arg->value,
+    [strip_size](std::string& s) {
+      s.erase(0, strip_size);
+    },
+    [](const auto&) {
+      TENZIR_UNREACHABLE();
+    });
+}
+
+auto to_located_string(ast::expression& expr) -> located<std::string> {
+  auto arg = try_as<ast::constant>(expr);
+  auto str = match(
+    arg->value,
+    [](std::string& s) -> std::string {
+      return std::move(s);
+    },
+    [](const auto&) -> std::string {
+      TENZIR_UNREACHABLE();
+      return {};
+    });
+  return located{std::move(str), arg->source};
+}
+
+auto strip_prefix(std::string name) -> std::string {
+  constexpr auto prefix = std::string_view{"tql2."};
+  if (name.starts_with(prefix)) {
+    return std::move(name).substr(prefix.size());
+  }
+  return name;
+}
+
+auto get_file(const boost::urls::url_view& url) -> std::string {
+  if (not url.segments().empty()) {
+    return url.segments().back();
+  }
+  if (url.host_type() == boost::urls::host_type::name) {
+    return url.host();
+  }
+  return {};
+}
+
+class from_plugin2 final : public virtual operator_factory_plugin {
+  static auto
+  create_pipeline_from_uri(std::string path, invocation inv,
+                           session ctx) -> failure_or<operator_ptr> {
```

Comment:
I put the implementation of `create_pipeline_from_uri` of `from_plugin2` and `to_plugin2` side-by-side and noticed that they are almost identical. Do you think that it makes sense from a maintainability perspective to already factor this out? Otherwise we always have to fix everything twice and minor divergence will eventually double the work.